### PR TITLE
[wip]Add volume control using scroll wheel

### PR DESCRIPTION
**Pull request by Myzel394**
**This Pull Request is not ready to be merged yet.**

## Fixes

Issue Number:

#1469 

## What is the current behavior?
No volume control using 

## What is the new behavior?
You can control the volume by scrolling your mouse wheel over the video.

## PR Checklist

<details><summary>Toggle...</summary>

What kind of change does this PR introduce?

- [ ] Bugfix
- [ X ] Feature
- [ ] Code style update (formatting)
- [ ] Refactoring (no functional changes)
- [ ] Documentation changes
- [ ] Other - Please describe:

Please check all that apply to this PR using "x":

- [ X ] I have checked that this PR is not a duplicate of an existing PR (open, closed or merged)
- [ ] I have checked that this PR does not introduce a breaking change
- [ ] This PR introduces breaking changes and I have provided a detailed explanation below

</details>

I don't know how to change the volume. Unfortunately, the code is honestly pretty messy and I don't want to spend hours just reverse-engineering how to do these basic things. You either have to provide some documentation or - even better - improve the code (why is there TypeScript code in a Javascript file for example?).
This is not meant to be toxic, but you need to understand that outsiders can't really work with the given code.
I wanted to contribute anyway, so here's what I did so far, hopefully it'll help you out.

### CODE REVIEW

**Comment by tzarebczan** [comment]
Thank you for the PR and feedback, appreciate it. We'll get it reviewed soon, but you'll also want to clean up the console logs.

**Comment by infinite-persistence** [comment]
tldr:  I think volume-scrolling should be confined to just the volume button/slider area (as suggested in 1469), or not implemented at all.

----

Cons:
- The behavior would be suitable for a desktop video player, but probably not on a webpage video player.  The mouse-wheel event should/needs to be scrolling the page instead, because:
  - When in Theater mode, it will be hard(er) to scroll the page as the video covers the majority of the screen.
  - The same goes for when browser width small and causes it to be in Mobile Layout.
    - Rafael, most likely not affected, but also need to check if it interferes with the touch gestures in real mobile.
- Blocks future enhancements what would require the mouse-event to not be tied to volume.

Suggestion to limit the event to the control bar's volume/slider area only for now (since that's what's being requested), or not implemented at all.

**Comment by tzarebczan** [comment]
@Myzel394 let us know if you wanna take a shot at changes to make the above work. Eventually we can add a setting to make this function like you original wanted also.

**Comment by Myzel394** [comment]
> @Myzel394 let us know if you wanna take a shot at changes to make the above work. Eventually we can add a setting to make this function like you original wanted also.

Yeah sure, but I don't know how to change the volume. Could you (or someone else) give me some details how to change the volume in the `VideoJS` component?

**Comment by Myzel394** [comment]
@tzarebczan Yes I'd like to work on this. However, I'm currently quite busy and like I already said it's a little difficult to add new code. I'm currently struggling how to properly use `dispatch`.

**Comment by Myzel394** [comment]
@tzarebczan Okay I found a way to add this functionality. There was a file called `videojs-keyboard-shortcuts.jsx` where I added this scroll shortcut.

You can change the volume by scrolling with the mouse over the video player. Change the volume precisely by holding the "SHIFT"-key while scrolling.

**Comment by mayeaux** [comment]
Hey @Myzel394 , thanks for the PR. This is looking great, just tested it out locally and it works really well. Something that came out of our discussion though is that probably to trigger this volume change if the cursor is anywhere over the video has a bit too much impact on user experience (it's very much expected and a norm on video sites that if you scroll down with your cursor on the video you will move your position in the document), that said are you able to have this be triggered when the cursor is hovering over the volume (either the icon or the horizontal bar) ? In that case I think we are happy to get this approved and merged, thanks for the contribution!

**Comment by mayeaux** [comment]
@Myzel394 I think that's a fine idea, I'd say set that up and I can try it out and see how it feels but I don't think it could realistically negatively impact the user experience, I've never heard of a shift scroll behaviour lol so that should be a safe behaviour.

**Comment by mayeaux** [comment]
Looks perfect! @Myzel394
Screen Shot 2022-05-27 at 12 27 22 PM
 
Screen Shot 2022-05-27 at 12 27 11 PM

Did run into this error while switching between videos, but overall the behaviour worked perfectly and adds a really nice touch, I think once this error gets fixed we can take a look at getting it merged ASAP, thanks for the contribution!

**Comment by Myzel394** [comment]
@mayeaux Should be fixed now

**Comment by mayeaux** [comment]
Going to get this checked out/merged ASAP

**Comment by tzarebczan** [comment]
@mayeaux can you fix the conflicts/linting and merge?

File: ui/component/viewers/videoViewer/internal/videojs.jsx [hunk 1]
```diff
@@ -431,7 +455,17 @@ export default React.memo<Props>(function VideoJs(props: Props) {
...
-      window.removeEventListener('keydown', curried_function);
+      const volumePanel = playerRef.current
+        .getChild(VIDEOJS_CONTROL_BAR_CLASS)
+        .getChild(VIDEOJS_VOLUME_PANEL_CLASS)
+        .el();
+
+      window.removeEventListener('keydown', keyDownHandlerRef.current);
+      containerRef.current.removeEventListener('wheel', videoScrollHandlerRef.current);
+
+      if (volumePanel.current) {
```
Comment (infinite-persistence) [inline ui/component/viewers/videoViewer/internal/videojs.jsx:466]: `volumePanel.current` doesn't make sense, because it's not a React `ref` but a DOM element.

I think this just masks the crash by never calling 'remove' code?

File: ui/component/viewers/videoViewer/internal/videojs.jsx [hunk 2]
```diff
@@ -431,7 +455,17 @@ export default React.memo<Props>(function VideoJs(props: Props) {
...
     // Cleanup
     return () => {
-      window.removeEventListener('keydown', curried_function);
+      const volumePanel = playerRef.current
+        .getChild(VIDEOJS_CONTROL_BAR_CLASS)
+        .getChild(VIDEOJS_VOLUME_PANEL_CLASS)
+        .el();
+
+      window.removeEventListener('keydown', keyDownHandlerRef.current);
+      containerRef.current.removeEventListener('wheel', videoScrollHandlerRef.current);
```
Comment (infinite-persistence) [inline ui/component/viewers/videoViewer/internal/videojs.jsx:464]: image

Just FYI for now -- can leave as-is, because ...

We plan to move the keyboard shortcut code into a plugin so that we don't need to inject these kind of changes here, which has nothing to do with the effect and violates the 1-thing-per-effect general rule. 

Inside the plugin, we can use the vjs framework to listen to these events instead of attaching the listeners ourselves.
```
player.on('mousewheel', mouseScroll);
componentXYZ.on('mousewheel', mouseScroll);
```

But can stay with this approach until then...

File: ui/component/viewers/videoViewer/internal/videojs.jsx [hunk 3]
```diff
@@ -371,7 +381,21 @@ export default React.memo<Props>(function VideoJs(props: Props) {
       // Set reference in component state
       playerRef.current = vjsPlayer;
 
-      window.addEventListener('keydown', curried_function(playerRef, containerRef));
+      const volumePanel = playerRef.current
+        .getChild(VIDEOJS_CONTROL_BAR_CLASS)
+        .getChild(VIDEOJS_VOLUME_PANEL_CLASS)
+        .el();
```
Comment (infinite-persistence) [inline ui/component/viewers/videoViewer/internal/videojs.jsx:387]: Fix lint please

**Review by infinite-persistence — CHANGES_REQUESTED** [review]
- [ ] squash and rebase to latest `master`? 
This is based off 2021 code ... while there are no direct conflicts, there might be behavioral conflicts (there are none for this specific case, but you get the idea).

**Comment by tzarebczan** [comment]
New PR: https://github.com/OdyseeTeam/odysee-frontend/pull/1638

**Comment by mayeaux** [comment]
Thanks @Myzel394 for the great contribution, much appreciated!

**Comment by Myzel394** [comment]
@mayeaux  Oh yeah Mister Krabs - thanks for merging it!
